### PR TITLE
integrate sequencer in Repairnator

### PR DESCRIPTION
**Pull request by jianguda**
#844

### CODE REVIEW

**Comment by monperrus** [comment]
thanks a lot.

what are the blockers for merging and the next steps?

**Comment by jaxxgu** [comment]
The main issue is that I can not perfectly build one SequenceR docker image. Therefore, I always get none patches in invoking SequenceR. More preciously, some Python libs can not be installed normally, I am trying to fix it. Once this issue get solved, I will complete test file and try docker-in-docker in CI.

**Comment by monperrus** [comment]
ok, thanks for the update. --Martin

**Comment by jaxxgu** [comment]
trying to solve issues which only occurred when using openJDK

**Comment by monperrus** [comment]
which version of openjdk?

**Comment by jaxxgu** [comment]
> which version of openjdk?

Repairnator is using openJDK8 while Astor is using oracleJDK8.

**Comment by jaxxgu** [comment]
I am comparing logs of different JDK env to help locate issue, here is my finding.

__openJDK__
```[INFO ] fr.inria.astor.core.faultlocalization.gzoltar.GZoltarFaultLocalization.searchSuspicious(GZoltarFaultLocalization.java:226) - Gzoltar Test Result Total:4, fails: 1, GZoltar suspicious 0, with positive susp 0```

__oracleJDK__
```[INFO ] fr.inria.astor.core.faultlocalization.gzoltar.GZoltarFaultLocalization.searchSuspicious(GZoltarFaultLocalization.java:226) - Gzoltar Test Result Total:4, fails: 1, GZoltar suspicious 8, with positive susp 3```

Maybe we could upgrade the __`gzoltar`__ lib from one ancient version to its latest version.

**Comment by monperrus** [comment]
See https://github.com/SpoonLabs/astor/pull/182

**Comment by monperrus** [comment]
> Maybe we could upgrade the gzoltar lib from one ancient version to its latest version.

Where? In Astor?

**Comment by jaxxgu** [comment]
> Maybe we could upgrade the gzoltar lib from one ancient version to its latest version.
> Where? In Astor?

Yes, in Astor. Besides, `Coming` is also using oracleJDK.

**Comment by jianguda** [comment]
waiting new `Astor` (with latest `gzoltar` lib)

**Comment by jianguda** [comment]
I plan to make master-branch's CI build okay firstly, then rebase the "issue844" branch. In this way I can have one stable PR. The reason is that CI build usually fail before `TestSequencerRepair`. It should relate to the change of repo's owner (as the CI environment seem also have changed).

**Comment by monperrus** [comment]
> I plan to make master-branch's CI build okay firstly

excellent idea

**Comment by monperrus** [comment]
you can rebase, master contains latest Astor

**Comment by jianguda** [comment]
For CI build error, the cause should be that, `Astor`'s [`getVersionJDK()`](https://github.com/SpoonLabs/astor/blob/master/src/main/java/fr/inria/astor/core/setup/ProjectConfiguration.java#L258) can not retrieve java version normally.

Here is red colored messages in my local log:
```
[ERROR] fr.inria.astor.core.setup.ProjectConfiguration.getVersionJDK(ProjectConfiguration.java:289) - Error retrieving java version: java.lang.NullPointerException
[ERROR] fr.inria.astor.core.setup.ProjectConfiguration.getVersionJDK(ProjectConfiguration.java:291) - Error retrieving java version output obtained:

java.lang.NullPointerException
	at java.io.File.<init>(File.java:277)
	at fr.inria.main.AbstractMain.processArguments(AbstractMain.java:422)
	at fr.inria.main.evolution.AstorMain.execute(AstorMain.java:214)
	at fr.inria.spirals.repairnator.process.step.repair.sequencer.SequencerRepair.businessExecute(SequencerRepair.java:77)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:334)
	at fr.inria.spirals.repairnator.process.step.repair.AbstractRepairStep.execute(AbstractRepairStep.java:52)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.repair.TestSequencerRepair.testSequencerRepair(TestSequencerRepair.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)


java.lang.NullPointerException
	at fr.inria.spirals.repairnator.process.step.repair.sequencer.SequencerRepair.businessExecute(SequencerRepair.java:83)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:334)
	at fr.inria.spirals.repairnator.process.step.repair.AbstractRepairStep.execute(AbstractRepairStep.java:52)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.executeNextStep(AbstractStep.java:187)
	at fr.inria.spirals.repairnator.process.step.AbstractStep.execute(AbstractStep.java:351)
	at fr.inria.spirals.repairnator.process.step.repair.TestSequencerRepair.testSequencerRepair(TestSequencerRepair.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

This is corresponding messages in [the failed CI build log](https://api.travis-ci.org/v3/job/583116827/log.txt):
```
Tests in error: 
  TestSequencerRepair.testSequencerRepair:79 Â» NullPointer
```

`TestSequencerRepair` is okay when run with oracleJDK8, but failed when with openJDK8. To help debug, in case `getVersionJDK()` exists issue, when I check `java -version`:
```
➜  ~ java -version                      
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_222-b10)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.222-b10, mixed mode)
```

**Comment by jianguda** [comment]
I ran `TestAstorRepair` to support my guess, and then also get same error messages:
```
[ERROR] fr.inria.astor.core.setup.ProjectConfiguration.getVersionJDK(ProjectConfiguration.java:289) - Error retrieving java version: java.lang.NullPointerException
[ERROR] fr.inria.astor.core.setup.ProjectConfiguration.getVersionJDK(ProjectConfiguration.java:291) - Error retrieving java version output obtained:

```

I find one strange thing that, we already have one successful CI build at master branch, where `TestAstorRepair` is okay, so `Astor`'s [`getVersionJDK()`](https://github.com/SpoonLabs/astor/blob/master/src/main/java/fr/inria/astor/core/setup/ProjectConfiguration.java#L258) should also be okay.

**Comment by jianguda** [comment]
Hey @monperrus I want some help, as I spent several days but still can not locate its cause.

In case someone needs info about how we use SequenceR docker image locally, you just need to run `docker pull ycaxgjd/sequencer`, or you can also run `docker pull zimin/sequencer`. The former is third-party image but has been tested, the latter is official image but has not been tested.

**Comment by monperrus** [comment]
Good diagnostic. In `getVersionJDK(String jvmPath)` what's the null variable triggering the NPE?

**Comment by jianguda** [comment]
> Good diagnostic. In `getVersionJDK(String jvmPath)` what's the null variable triggering the NPE?

I am not sure, as it is not very convenient to debug one lib, also I do not know error trace. However, I feel it is most likely that, we get `null` from `ConfigurationProperties.getProperty("location")` [here](https://github.com/SpoonLabs/astor/blob/master/src/main/java/fr/inria/astor/core/setup/ProjectConfiguration.java#L274).

To ensure we won't get `null` from `ConfigurationProperties.getProperty("location")`, I checked [AbstractMain.java#L395](https://github.com/SpoonLabs/astor/blob/master/src/main/java/fr/inria/main/AbstractMain.java#L395), then I found `ProjectConfiguration.getVersionJDK(jvmhome)` occurred [AbstractMain.java#L377](https://github.com/SpoonLabs/astor/blob/master/src/main/java/fr/inria/main/AbstractMain.java#L377). So I understand why NPE, it is because the location property is required before assigned.

To solve this issue, we could move L372-384 to the end of this function. But I am not sure whether I am right, because `TestAstorRepair` won't trigger this NPE. It is really confusing.

**Comment by monperrus** [comment]
> I am not sure, as it is not very convenient to debug one lib

Yes. The solution is to change Astor locally, to make `mvn install` and to change the version in
Repairnator's pom.

Is there a workaround, eg to set the location before calling Astor/Sequencer?

**Comment by monperrus** [comment]
FYI https://github.com/SpoonLabs/astor/pull/188

**Comment by jianguda** [comment]
> Is there a workaround, eg to set the location before calling Astor/Sequencer?

We can move L372-384 to [AbstractMain.java#L406](https://github.com/SpoonLabs/astor/blob/master/src/main/java/fr/inria/main/AbstractMain.java#L406)

**Comment by jianguda** [comment]
> FYI [SpoonLabs/astor#188](https://github.com/SpoonLabs/astor/pull/188)

Could we have one new version of Astor? Then I can check whether this NPE is real cause.

**Comment by jianguda** [comment]
> Yes. The solution is to change Astor locally, to make `mvn install` and to change the version in Repairnator's pom.

I tried both your fix and my fix to this NPE, and found this NPE is not real cause.

I am trying to find the real cause.

**Comment by monperrus** [comment]
ok, thanks.

**Comment by jianguda** [comment]
Hey @monperrus
I finally solved the Astor's issue, FYI [SpoonLabs/astor#189](https://github.com/SpoonLabs/astor/pull/189).

I have already passed my test locally, and I will try to pass CI once latest `Astor` is prepared.

**Comment by monperrus** [comment]
Congrats, thanks a lot for the debugging effort!

**Comment by jianguda** [comment]
Hey @monperrus 
This branch is ready to merge.
(It is strange that it still fails even though I have "Signed-off-by footer")

File: .travis/travis-before.sh [hunk 1]
```diff
@@ -5,6 +5,7 @@ sudo apt-get install -y nodejs
 sudo npm install -g ajv-cli
 sudo apt-get install shunit2
 docker pull repairnator/pipeline
+docker pull ycaxgjd/sequencer:1.0
```
Comment (monperrus) [inline .travis/travis-before.sh]: push to docker pull repairnator/sequencer:1.0

File: repairnator/docker-images/pipeline-dockerimage/pipeline_launcher.sh [hunk 2]
```diff
@@ -82,3 +82,9 @@ export GITHUB_USEREMAIL=
 
 echo "Execute pipeline with following supplementary args: $args"
 java -cp $JAVA_HOME/lib/tools.jar:repairnator-pipeline.jar -Dlogback.configurationFile=/root/logback.xml fr.inria.spirals.repairnator.pipeline.Launcher -d -b $LOCAL_BUILD_ID --runId $LOCAL_RUN_ID -o $LOCAL_OUTPUT --ghOauth $LOCAL_GITHUB_OAUTH --repairTools $REPAIR_TOOLS $args
+
+# prepare the SequenceR docker image
+# https://github.com/KTH/sequencer
+# https://hub.docker.com/r/zimin/sequencer
+# https://hub.docker.com/r/ycaxgjd/sequencer
+docker pull ycaxgjd/sequencer:1.0
```
Comment (monperrus) [inline repairnator/docker-images/pipeline-dockerimage/pipeline_launcher.sh]: same

File: repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerRepair.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    @Override
+    public String getRepairToolName() {
+        return TOOL_NAME;
+    }
+
+    @Override
+    protected StepStatus businessExecute() {
+        this.getLogger().info("Start SequencerRepair");
+        String pathPrefix = ""; // for macOS: "/private";
+        String imageTag = "ycaxgjd/sequencer:1.0";
```
Comment (monperrus) [inline repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerRepair.java]: same

File: repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerRepair.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringJoiner;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+public class SequencerRepair extends AbstractRepairStep {
```
Comment (monperrus) [inline repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerRepair.java]: short javadoc?

File: repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerResult.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import com.google.common.io.Files;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class SequencerResult {
```
Comment (monperrus) [inline repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerResult.java]: short javadoc?

File: .travis/travis-before.sh (hunk 1 above)
Comment (monperrus) [reply .travis/travis-before.sh]: I gave you the rights on DockerHub

Comment (jianguda) [reply .travis/travis-before.sh]: > I gave you the rights on DockerHub

Hey @martinezmatias Could you please check my permissions?
As when I push, I always meet `denied: requested access to the resource is denied`.

Comment (jianguda) [reply .travis/travis-before.sh]: Hey @monperrus 
Maybe I only have the read permission.

File: repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerRepair.java (hunk 4 above)
Comment (jianguda) [reply repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerRepair.java]: > short javadoc?

added

File: repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerResult.java (hunk 5 above)
Comment (jianguda) [reply repairnator/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/sequencer/SequencerResult.java]: > short javadoc?

added

**Comment by monperrus** [comment]
Fixed, could you try again?

**Comment by jianguda** [comment]
> Fixed, could you try again?

I have done.

**Comment by jianguda** [comment]
As this PR has been reviewed by @monperrus , I merge it by myself.

**Pull request merged**